### ISSUE TITLE
Glassfish start-domain fails. DOIEZIdServer problem?

### ISSUE THREAD

**Issue by lmaylein** [IQSS/dataverse#4903]
Suddenly (since this morning) we are no longer able to start our dataverse instance.

```
/etc/init.d/glassfish start
Starting GlassFish server: glassfishEnter master password (3) attempt(s) remain)>
Waiting for domain1 to start .......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
serverNoStart
Command start-domain failed.
```

The admin console (Port 4848) shows:
`Status: The Admin Console is starting. Please wait.`

After that, I have to use `kill 9 <pid>` to stop the process.

The last message in the server.log is

[2018-07-27T09:55:34.579+0200] [glassfish 4.1] [INFO] [AS-EJB-00054] [javax.enterprise.ejb.container] [tid: _ThreadID=18 _ThreadName=RunLevelControllerThread-1532678107499] [timeMillis: 1532678134579] [levelValue: 800] [[
  Portable JNDI names for EJB DOIEZIdServiceBean: [java:global/dataverse-4.8.6/DOIEZIdServiceBean!edu.harvard.iq.dataverse.DOIEZIdServiceBean, java:global/dataverse-4.8.6/DOIEZIdServiceBean]]]

netstat shows an open connection to  34.250.137.166:443 (Amazon Cloud)

Is it possible that the startup process is blocked by an EZID connection problem.
Please note: Our dataverse is configured to use Datacite!

Our dataverse version: 4.8.6

**Comment by lmaylein** [comment]
Ok. Obviously DOIEZIdServiceBean is also responsive for datacite interface. The mentioned ip is part of the datacite API.

At the moment 
http://status.datacite.org/
shows an "Metadata Store (MDS) API" outage of 23 minutes but our problem has been around for at least 2 hours.

**Comment by donsizemore** [comment]
I'm not part of IQSS, but am up early, and in case this helps:

- with Glassfish stopped, rm -rf /usr/local/glassfish4/glassfish/domains/domain1/generated (or equivalent)
- also with Glassfish stopped, hop in Postgres and remove any entries from the "EJB__TIMER__TBL" table

Then try to start Glassfish? This may be a result of #4902 - on our production system, we have 64GB of VM RAM and set an Xmx value of 32768m.

**Comment by lmaylein** [comment]
... and https://mds.datacite.org/ is not responding :-(

**Comment by pdurbin** [comment]
I have a server at https://dev1.dataverse.org configured to use DataCite that's running 7608db9 which is some code that recently got merged into the "develop" branch. So it's pretty much 4.9.1 plus the stuff we've been working on for 4.9.2.

I just did a little experiment to see if I can replicate this problem but I wasn't able to.

That server is configure to use the test MDS service at https://mds.test.datacite.org which seems to be working fine. I was able to publish a dataset.

I switched it over to the production MDS service at https://mds.datacite.org and attempted to publish a dataset. As expected, this failed and server.log reported `Response code: 401, Bad credentials` because I only have a username and password for the test MDS, not the production MDS.

Then I restarted Glassfish while I still had Dataverse configured to use the production MDS. Dataverse started just fine.

It would seem odd to me that an external service like MDS being down would cause Dataverse not to start. However, I'm not sure what the solution for you is. I think @donsizemore is probably on the right track that the out of memory error may have left something behind but I don't have any ideas beyond deleting that "generated" folder (which I do all the time in development, by the way).

@lmaylein I guess I would suggested emailing your server.log file to support@dataverse.org just after you attempt to start Dataverse (so that we can see the most recent logging leading to the problem at the end).

**Comment by lmaylein** [comment]
Ok. I'll send you the server.log
Could you please check, if you could reach https://mds.datacite.org ? Maybe it's a regional problem.

**Comment by pdurbin** [comment]
Wow, that's crazy. I wonder why the opposite wasn't true in my test. Fingers cross that when https://mds.datacite.org comes back up you are able to start Glassfish. @mfenner do you have any information on the status other than what we see at http://status.datacite.org ?

**Comment by lmaylein** [comment]
Maybe https://mds.datacite.org is reachable from your location. It seems to me that the datacite service resides in amazon cloud. Therefore there could be differences in reachability.

**Comment by pdurbin** [comment]
@lmaylein no, https://mds.datacite.org is definitely down for me too. When I was doing my test I saw "504 Gateway Time-out" in my browser. And testing right now from that dev1 server itself, I see the same:

```
[pdurbin@dvnweb-vm2 ~]$ curl https://mds.datacite.org
<html>
<head><title>504 Gateway Time-out</title></head>
<body bgcolor="white">
<center><h1>504 Gateway Time-out</h1></center>
</body>
</html>
[pdurbin@dvnweb-vm2 ~]$ 
```

So why is it that Dataverse starts when my server is configured to use https://mds.datacite.org but yours does not? It's down for both of us.

Meanwhile I don't see anything particularly helpful in the logs you sent at https://help.hmdc.harvard.edu/Ticket/Display.html?id=264966 (a couple ticket numbers got merged together) but thank you for sending them.

Have you thought about what your support options might be from the Glassfish side? Maybe there's a mailing list or something? No one has responded to me in #glassfish on freenode IRC yet.

**Comment by lmaylein** [comment]
I don't think it's a glassfish problem. There could be differences in connection management between different java versions or operating systems...

Could you see something in our logs which causes the problems described in #4902

**Comment by pdurbin** [comment]
@lmaylein I'm looking at the logs but I'll leave a comment about the OutOfMemoryError in #4902

Did you notice that @donsizemore asked about EJB__TIMER__TBL at https://github.com/IQSS/dataverse/issues/4903#issuecomment-408375445 ? There's some more information on this at http://guides.dataverse.org/en/4.9.1/admin/troubleshooting.html but I'm not sure if it will help you get your Dataverse installation back up or not.

**Comment by pameyer** [comment]
Also not IQSS - but under the assumption that failure to access https://mds.datacite.org is the source of the problem (I'm seeing 504 from that URL as well), one option to try might be to use /etc/hosts to switch mds.datacite.org to mds.test.datacite.org (where I'm seeing 200) and see if glassfish starts that way.  Obviously, this is something you'd want to switch back after glassfish startup (and won't help with any actual publish/edit metadata/etc calls to production datacite.

**Comment by lmaylein** [comment]
@pameyer Thanks. To change /etc/hosts is probably a better idea than to change the url in domain.xml  (my current workaround).

**Comment by pdurbin** [comment]
@pameyer good though. @lmaylein already established that he **can** start Glassfish with the test MDS: https://github.com/IQSS/dataverse/issues/4903#issuecomment-408391710

What you and he may not know is that after Glassfish has been started, you can reconfigure from test MDS to production MDS on the fly with asadmin commands so that you don't have to restart Glassfish. http://guides.dataverse.org/en/4.9.1/installation/config.html#doi-baseurlstring has the following example:

Out of the box, Dataverse is configured to use base URL string from EZID. You can delete it like this:

./asadmin delete-jvm-options '-Ddoi.baseurlstring=https\://ezid.cdlib.org'

Then, to switch to DataCite, you can issue the following command:

./asadmin create-jvm-options '-Ddoi.baseurlstring=https\://mds.datacite.org'

You would need to replace these with the appropriate values, of course.

**Comment by djbrooke** [comment]
Hey @lmaylein - do you have any updates on this? Is there any change in status or is there anything else you've tried that may be helpful for us to know?

**Comment by lmaylein** [comment]
MDS is now online again. At the moment our Dataverse starts and stops without any problems.

**Comment by mfenner** [comment]
We at DataCite have an issue with handle registration that is intermittently not working. A permanent fix is in the works and will be deployed after more testing tomorrow.

**Comment by pdurbin** [comment]
@mfenner thanks!

@lmaylein any ideas on how a developer can reproduce this problem? I'm glad to hear you're back up now. It's still crazy to me that Glassfish won't start for you when MDS is down. I have no idea why. 😕

**Comment by lmaylein** [comment]
Unfortunately I'm not able to reproduce this problem.

Blocking outgoing https requests (`iptables -A OUTPUT -p tcp -d 0.0.0.0/0.0.0.0 --dport 443 -j DROP`) only delays glassfish start.

```
[2018-07-30T12:53:53.909+0200] [glassfish 4.1] [INFO] [AS-EJB-00054] [javax.ente
rprise.ejb.container] [tid: _ThreadID=18 _ThreadName=RunLevelControllerThread-15
32948009760] [timeMillis: 1532948033909] [levelValue: 800] [[
  Portable JNDI names for EJB DOIEZIdServiceBean: [java:global/dataverse-4.8.6/D
OIEZIdServiceBean!edu.harvard.iq.dataverse.DOIEZIdServiceBean, java:global/datav
erse-4.8.6/DOIEZIdServiceBean]]]

[2018-07-30T12:58:09.148+0200] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.
core.index.DOIEZIdServiceBean] [tid: _ThreadID=18 _ThreadName=RunLevelController
Thread-1532948009760] [timeMillis: 1532948289148] [levelValue: 900] [[
  login failed ]]

[2018-07-30T12:58:09.149+0200] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.
core.index.DOIEZIdServiceBean] [tid: _ThreadID=18 _ThreadName=RunLevelController
Thread-1532948009760] [timeMillis: 1532948289149] [levelValue: 900] [[
  Exception String: edu.ucsb.nceas.ezid.EZIDException: Connection timed out (Con
nection timed out)]]

[2018-07-30T12:58:09.151+0200] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.
core.index.DOIEZIdServiceBean] [tid: _ThreadID=18 _ThreadName=RunLevelController
Thread-1532948009760] [timeMillis: 1532948289151] [levelValue: 900] [[
  localized message: Connection timed out (Connection timed out)]]

[2018-07-30T12:58:09.152+0200] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.
core.index.DOIEZIdServiceBean] [tid: _ThreadID=18 _ThreadName=RunLevelController
Thread-1532948009760] [timeMillis: 1532948289152] [levelValue: 900] [[
  cause: ]]

[2018-07-30T12:58:09.152+0200] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.
core.index.DOIEZIdServiceBean] [tid: _ThreadID=18 _ThreadName=RunLevelController
Thread-1532948009760] [timeMillis: 1532948289152] [levelValue: 900] [[
  message Connection timed out (Connection timed out)]]
```

I assume this problem only occurs if MDS is in the same state as it was three days ago (tcp connection okay but problems with the service itself).

**Comment by lmaylein** [comment]
I'm afraid, there is no alternative...

**Comment by pdurbin** [comment]
@lmaylein ok, closing for now. Thanks.
